src/editor/search/TextSearch.java

```diff
@@ -0,0 +1,113 @@
+package editor.search;
+
+import editor.ui.TextEditor;
+
+import java.util.ArrayList;
+
+/**
+ *  Base class for searching elements of text inside of it
+ *
+ * @see FindMatches -creates list of indexes of match(their coordinates in text)
+ * @see ShowPrevMatch -selects previous match
+ * @see ShowNextMatch -selects next match
+ */
+public class TextSearch {
+
+    /**
+     *  List of indexes of matches inside of text
+     */
+    static  ArrayList<Integer> startIndexes;
+    static  ArrayList<Integer> endIndexes;
+    /**
+     *  Current position inside of matches indexes
+     */
+    static int matchNum;
```

Comment:
There are some concurrency issues. Several threads work with this variable without any kind of synchronization.
If one thread writes a value to a variable, another thread sometimes cannot see the changes. So, some kind of synchronization is needed here. The same with lists.

I would prefer to create an instance TextSearch searcher and pass it to threads which should work with it. To synchronize access to the data you may use the synchronized keyword and special thread-safe collections from java sdk.

Reply:
I've tried to realise thread safety  in "search" method, where new threads not start until previous not finished. I'll read about thread-safety collections and will use synchronized keyword. Thanks for attention!